### ISSUE TITLE
Add schema validation for unknown keys in required property

### ISSUE THREAD

**Issue by ariesclark** [ExodusMovement/schemasafe#150]
```json
{
    "type": "object",
    "required": ["foo", "bar"],
    "properties": {
        "foo": {
            "type": "number"
        }
    }
}
```

I think the schemasafe *strict?* validator should throw an error in this case.

**Comment by ChALkeR** [comment]
Hm. Technically, this is not an error per the JSON Schema spec, and has a reasonable meaning.

Schema like that doesn't make sense if there is an immediate `additionalProperties: false` there though.

And using [strong mode](https://github.com/ExodusMovement/schemasafe/blob/master/doc/Strong-mode.md) of schemasafe, it will ensure that all properties should be validated.

But they could be validated via more complex rules than the one here, and e.g. doing this is possible even in strong mode:
```
{
  $schema: 'http://json-schema.org/draft/2020-12/schema#',
  type: 'object',
  required: ['foo', 'bar'],
  allOf: [
    {
      required: [],
      properties: {
        foo: {
          type: 'number',
        },
      },
    },
    {
      required: [],
      properties: {
        bar: {
          type: 'number',
        },
      },
    },
  ],
  unevaluatedProperties: false,
}
```
Also, a more complex example would be possible with refs.

That said, a validation of `required` in cases where we can statically prove that those properties are not possible can be done.
Just not with this example (as it's missing `additionalProperites` or `unevaluatedProperties`).

Sorry for the late reply.

**Comment by ChALkeR** [comment]
Implemented in 1007440914ba51ba0264c972c2dd455f63642c72.

Not that this check is present only for cases when either `additionalProperties: false` or a statically-provable `unevaluatedProperties: false` are used, so we can be sure that `required` attempts to list a property that is forbidden by either of those.

**Comment by ChALkeR** [comment]
This check also works with inherited `required`, i.e. this schema throws at compile time:

```js
{
  type: 'object',
  required: [],
  properties: {
    foo: {
      type: 'number',
    },
  },
  allOf: [
    {
      required: ['foo', 'bar'],
    },
  ],
  unevaluatedProperties: false,
}
```

That said, `additionalProperties: false` checks only adjacent `required` for now, so this would pass with `additionalProperties` instead of `unevaluatedProperties`.

Let's reopen as there is room for improvement.
